src/tokenizer.js

```diff
@@ -0,0 +1,801 @@
+// <semicolon-token>, <comma-token>, <[-token>, <]-token>, <(-token>, <)-token>, <{-token>, and <}-token>.
+export class IdentToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class FunctionToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class AtKeywordToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class HashToken extends Token {
+  type;
+  value;
+  constructor(value, type = 'unrestricted') {
+    super();
+    this.value = value;
+    this.type = type;
+  }
+}
+
+export class StringToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class BadStringToken extends Token {}
+
+export class UrlToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class BadUrlToken extends Token {}
+
+export class DelimToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class NumberToken extends Token {
+  value;
+  type;
+  constructor(value, type = "integer") {
+    super();
+    this.value = value;
+    this.type = type;
+  }
+}
+
+export class PercentageToken extends Token {
+  value;
+  constructor(value) {
+    super();
+    this.value = value;
+  }
+}
+
+export class DimensionToken extends Token {
+  value;
+  type;
+  unit;
+  constructor(value, type, unit) {
+    super();
+    this.value = value;
+    this.type = type;
+    this.unit = unit;
+  }
+}
+
+export class WhitespaceToken extends Token {}
+
+export class CDOToken extends Token {}
+
+export class CDCToken extends Token {}
+
+export class ColonToken extends Token {}
+
+export class SemicolonToken extends Token {}
+
+export class CommaToken extends Token {}
+
+export class LeftSquareBracketToken extends Token {}
+
+export class RightSquareBracketToken extends Token {}
+
+export class LeftParenthesisToken extends Token {}
+
+export class RightParenthesisToken extends Token {}
+
+export class LeftCurlyBracketToken extends Token {}
+
+export class RightCurlyBracketToken extends Token {}
+
+class InputStream {
+  input
+  index = 0;
+  constructor(input) {
+    this.input = input;
+  }
+
+  consume() {
+    const codePoint = this.input.codePointAt(this.index);
+    if (typeof codePoint !== 'undefined') {
+      this.index += String.fromCodePoint(codePoint).length;
+    }
+    return codePoint;
+  }
+
+  reconsume(codePoint) {
+    if (typeof codePoint !== 'undefined') {
+      this.index -= String.fromCodePoint(codePoint).length
+    }
+  }
+
+  peek() {
+    const codePoints = []
+    let position = this.index
+    for (let i = 0; i < 3; i++) {
+      const nextCodePoint = this.input.codePointAt(position);
+      if (typeof nextCodePoint !== 'undefined') {
```

Comment:
According to the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt), codePointAt only returns undefined if the index is out of range. I think it would be cleaner to change the for loop condition to `i < 3 && position < this.input.length`, then if the loop runs we always expect it to find a codepoint.

Reply:
Yeah.. That looks cleaner: 
Fixed here: https://github.com/flackr/scroll-timeline/compare/8e7cf381430bd1d2c7511debec7d954ebaccaf66..4ed64d0898c93a488381f6ad35c42033b39698a8

Reply:
🙏